### PR TITLE
[loki-distributed] Added support for IPv6 only clusters

### PR DESCRIPTION
**Pull request by stevehipwell**
This PR enables the modification of the Loki Gateway Nginx listener without needing to duplicate and maintain the whole `gateway.nginxConfig.file` value; this functionality is required to use Loki in an EKS IPv6 cluster. Unless you change `gateway.nginxConfig.serverListens` this change should be a no-op.

### CODE REVIEW

**Comment by zanhsieh** [comment]
@stevehipwell 
Just curious, how did you install and test this in your local machine? I plan for rework on the CI build for this.

**Comment by stevehipwell** [comment]
@zanhsieh I tested it from my local machine.

**Comment by zanhsieh** [comment]
> @zanhsieh I tested it from my local machine.

I mean, how did you set up your local stack as IPv6 only?

**Comment by stevehipwell** [comment]
> I mean, how did you set up your local stack as IPv6 only?

I didn't, I'm not sure it's possible, I hit an IPv6 EKS cluster.

**Comment by stevehipwell** [comment]
@unguiculus @Whyeasy could one of you give this a review?

File: charts/loki-distributed/values.yaml
Comment (zanhsieh) [inline charts/loki-distributed/values.yaml:948]: I think we should make it clear that this stanza is for IPv6, and IP6Enabled to enable / disable it (default is false).

```suggestion
    IP6serverListens: |-
      listen             8080;

    # -- Enable IPv6 only clusters configure
    IP6Enabled: false
```

File: charts/loki-distributed/README.md
Comment (zanhsieh) [inline charts/loki-distributed/README.md:156]:
```suggestion
| gateway.nginxConfig.IP6serverListens | string | `"listen             8080;"` | Configure the server listener(s), this needs to be customised for IPv6 only clusters |
| gateway.nginxConfig.IP6Enabled | bool | `false` | Enable IPv6 only clusters configure |
```

**Review by zanhsieh — CHANGES_REQUESTED** [review]

File: charts/loki-distributed/values.yaml
Comment (stevehipwell) [reply charts/loki-distributed/values.yaml:948]: @zanhsieh the current use of this block is for IP family so I could get behind using `IpFamilyServerListens` but I really don't see the benefit of that form over `serverListens`. Could you explain in a bit more detail why you think we need to explicitly tie the block to the current implementation while the listens could be modified for reasons other than just switching IP family?

Comment (stevehipwell) [reply charts/loki-distributed/values.yaml:948]: @zanhsieh why don't we add more comments and leave it as it is, for testing purposes you can just prove that if the user leaves it alone it's a no-op? AFAIK the rest of the Nginx config is like this.

I'd have loved to be able to provide a higher level IP family argument to hide the complexity from the end users but this isn't possible as the Loki Distributed chart [won't start on IPv6 cluster](https://github.com/grafana/helm-charts/issues/157) without a hack (see https://github.com/grafana/dskit/pull/185 for the actual fix) and even when it's running it doesn't [actually work](https://github.com/grafana/loki/issues/5578).

**Comment by mossad-zika** [comment]
so?

**Comment by zanhsieh** [comment]
> so?

@mossad-zika 
If the original author can't provide any way to test this PR and take care backward compatibility, I won't let it pass. Other maintainer(s) might let it passed.

**Comment by stevehipwell** [comment]
@zanhsieh I did test the chart and the PR is a no-op so I'm confused about your attitude to the change.

**Comment by zalegrala** [comment]
It might be a good idea to wait for the dskit pr to be merged upstream, and then integrated into loki, just in case we decide to rename a config variable or some such.  Love the effort to test this out using the chart though, pretty nice.

**Comment by stevehipwell** [comment]
@zalegrala this PR is intended to allow the gateway to support IPv6 but is actually a generic change which results in a no-op unless utilised so it could me merged immediately which would allow easier testing of the dskit changes.

**Comment by zalegrala** [comment]
Oh nice @stevehipwell, makes sense to support ahead of the dskit pr.

**Comment by stevehipwell** [comment]
Could someone give this a review please?

**Pull request closed without merging**
